rlp: remove elastic-array crate and use Vec instead

tafia opened the pull request:
The main purpose is to reduce the unsafe surface (elastic-array uses a ton of them).
The unexpected consequence is to have a nice perf boost.

On my machine:
```
$ cargo benchcmp master vec 
 name                             master ns/iter  vec ns/iter  diff ns/iter   diff %  speedup 
 bench_decode_nested_empty_lists  400             368                   -32   -8.00%   x 1.09 
 bench_decode_u256_value          73              69                     -4   -5.48%   x 1.06 
 bench_decode_u64_value           58              56                     -2   -3.45%   x 1.04 
 bench_stream_1000_empty_lists    14,671          7,921              -6,750  -46.01%   x 1.85 
 bench_stream_nested_empty_lists  303             145                  -158  -52.15%   x 2.09 
 bench_stream_u256_value          440             354                   -86  -19.55%   x 1.24 
 bench_stream_u64_value           140             70                    -70  -50.00%   x 2.00 
```

My guess is, given that elastic-array is quite an old crate, that it is a combination of
- Vec implementation being better than it used to be
- compiler being smarter and probably working very well with Vec in particular

It can also be that the bench is somehow wrong or does not represent the reality. I am not an expert enough on your use case to tell.

Regarding formating issues, I have tried to use parity-ethereum rustfmt.toml but lot of config options are not recognized anymore by rustfmt-preview (stable 1.28). Do you have a recommendation here?

Oh, and this is a breaking change. I can adapt the other repos if needed later on.

tafia commented:
I got an error when trying to sign the CLA.

```
AuthorizationError: The redirect_uri MUST match the registered callback URL for this application. at Strategy.OAuth2Strategy.authenticate (/app/cla-bot/node_modules/passport-oauth2/lib/strategy.js:131:25) at attempt (/app/cla-bot/node_modules/passport/lib/middleware/authenticate.js:348:16) at authenticate (/app/cla-bot/node_modules/passport/lib/middleware/authenticate.js:349:7) at callbacks (/app/cla-bot/node_modules/express/lib/router/index.js:164:37) at param (/app/cla-bot/node_modules/express/lib/router/index.js:138:11) at pass (/app/cla-bot/node_modules/express/lib/router/index.js:145:5) at Router._dispatch (/app/cla-bot/node_modules/express/lib/router/index.js:173:5) at Object.router (/app/cla-bot/node_modules/express/lib/router/index.js:33:10) at next (/app/cla-bot/node_modules/connect/lib/proto.js:193:15) at Object.compress [as handle] (/app/cla-bot/node_modules/connect/lib/middleware/compress.js:182:5) at next (/app/cla-bot/node_modules/connect/lib/proto.js:193:15) at /app/cla-bot/node_modules/connect-file-cache/lib/cache.js:54:18 at ConnectFileCache.loadFile (/app/cla-bot/node_modules/connect-file-cache/lib/cache.js:63:16) at ConnectFileCache.middleware (/app/cla-bot/node_modules/connect-file-cache/lib/cache.js:48:19) at Object.handle (/app/cla-bot/node_modules/connect-file-cache/lib/cache.js:4:61) at next (/app/cla-bot/node_modules/connect/lib/proto.js:193:15) at SessionStrategy.strategy.pass (/app/cla-bot/node_modules/passport/lib/middleware/authenticate.js:325:9) at SessionStrategy.authenticate (/app/cla-bot/node_modules/passport/lib/strategies/session.js:71:10) at attempt (/app/cla-bot/node_modules/passport/lib/middleware/authenticate.js:348:16) at Object.authenticate [as handle] (/app/cla-bot/node_modules/passport/lib/middleware/authenticate.js:349:7) at next (/app/cla-bot/node_modules/connect/lib/proto.js:193:15) at Object.initialize [as handle] (/app/cla-bot/node_modules/passport/lib/middleware/initialize.js:53:5) at next (/app/cla-bot/node_modules/connect/lib/proto.js:193:15) at next (/app/cla-bot/node_modules/connect/lib/middleware/session.js:315:9) at /app/cla-bot/node_modules/connect/lib/middleware/session.js:339:9 at /app/cla-bot/node_modules/connect/lib/middleware/session/memory.js:50:9 at _combinedTickCallback (internal/process/next_tick.js:73:7) at process._tickCallback (internal/process/next_tick.js:104:9) 
```

dvdplm commented:
@tafia Thank you for your contribution, good stuff here. Can you make the rlp tests pass please? And we'll need a version bump for sure.

dvdplm commented:
FWIW here's the benchmarks on my machine:

```
$ cargo benchcmp master vec
 name                             master ns/iter  vec ns/iter  diff ns/iter   diff %  speedup
 bench_decode_nested_empty_lists  213             215                     2    0.94%   x 0.99
 bench_decode_u256_value          36              36                      0    0.00%   x 1.00
 bench_decode_u64_value           37              38                      1    2.70%   x 0.97
 bench_stream_1000_empty_lists    8,284           5,798              -2,486  -30.01%   x 1.43
 bench_stream_nested_empty_lists  183             100                   -83  -45.36%   x 1.83
 bench_stream_u256_value          271             237                   -34  -12.55%   x 1.14
 bench_stream_u64_value           85              50                    -35  -41.18%   x 1.70
```

tafia commented:
> Can you make the rlp tests pass please?

Yes. There are a lot of dependencies.
Do you prefer 
- a first pass where I convert all `Vec` back into `ElasticSearchArray`? It would then be a minor update for them.
- modify the whole thing so it won't unecessarily depend on ElasticSearch? This is a bigger change but it is kind of more consistent. There may be several breaking changes, thus requiring major version bump.


What to I do for the CLA?

dvdplm commented:
> There are a lot of dependencies.

A *lot*: this change breaks a bunch of stuff in `parity-common` but likely breaks a ton of stuff in other repos as well. Like a *ton*.

I'm not opposed to go ahead with this anyway, but it could take us quite a while to get to a point where we can merge this.

I'd like to hear from @debris what he thinks. Possibly a better way would be to change `elastic-array` to just be a thin wrapper around `Vec`?

As for the CLA, I've pinged some people! Working on it! :)

ordian commented:
There are also concerns about cache locality and allocator traffic, which `elastic-array` is supposed to improve. These benchmarks may be not very representative. Also I'm curious to see how `elastic-array` holds against recent version of `smallvec` (cf https://github.com/paritytech/parity-ethereum/pull/4874).

tafia commented:
@dvdplm I can probably help fix all repos.
@ordian yes this is definitely possible. I think I'll try to go ahead anyway and see what other benches say.

dvdplm commented:
@tafia Wow, that's a generous offer! A good place to start would be the benches in `patricia_trie` I think.

ordian commented:
> Regarding formating issues, I have tried to use parity-ethereum rustfmt.toml but lot of config options are not recognized anymore by rustfmt-preview (stable 1.28). Do you have a recommendation here?

Here is our [style guide](https://github.com/paritytech/polkadot/wiki/Style-Guide).

dvdplm commented:
[CLA issue](https://github.com/paritytech/devops/issues/177)

ordian commented:
@tafia please try signing CLA again:  https://cla-parity-common.parity.io

tafia commented:
First sorry for all the posts.

After updating NodeCodec to use Vec too, I got slightly more consistent results:
```
$ cargo benchcmp master vec
 name                        master ns/iter  vec ns/iter  diff ns/iter  diff %  speedup 
 trie_insertions_32_mir_1k   3,531,682       3,509,978         -21,704  -0.61%   x 1.01 
 trie_insertions_32_ran_1k   3,495,609       3,534,305          38,696   1.11%   x 0.99 
 trie_insertions_random_mid  2,661,568       2,588,041         -73,527  -2.76%   x 1.03 
 trie_insertions_six_high    2,694,820       2,574,611        -120,209  -4.46%   x 1.05 
 trie_insertions_six_low     5,148,732       5,242,602          93,870   1.82%   x 0.98 
 trie_insertions_six_mid     3,453,874       3,393,769         -60,105  -1.74%   x 1.02 
 trie_iter                   2,631,114       2,519,411        -111,703  -4.25%   x 1.04
```

dvdplm commented:
@tafia interesting! What happens if you update `hashdb::DBValue` to use `Vec`? It's a type that is used all over and might have an impact too. There's also `triehash` that has benchmarks but I don't think it'd have much of an impact.

tafia commented:
> @tafia interesting! What happens if you update hashdb::DBValue to use Vec

I am in the process of changing `hasdb::DBValue` and friends (memoryDb etc ...) to be generic over the type of value they use (which means eliminate `DBValue` for a generic `T`).

I figured it was actually simpler than changing everything so we can decide on the best strategy case by case (either keeping the ElasticArray or try using a Vec ... or even use a bare array). I'll see where it goes, as you can expect this will be a much bigger PR than anticipated.

gabreal commented:
@tafia I don't know what happend. It's working for me now, can you maybe please try once more :)
https://cla-parity-common.parity.io

dvdplm commented:
> this will be a much bigger PR than anticipated.

Yeah. Keep in mind that many of these types are used in e.g. [`parity-ethereum`](https://github.com/paritytech/parity-ethereum/search?p=1&q=DBValue&unscoped_q=DBValue) and [`substrate`](https://github.com/paritytech/substrate/search?q=DBValue&unscoped_q=DBValue) too. So yeah, definitely a lot of grunt work to make this happen!

tafia commented:
@gabreal it worked this time. Not sure what happened (did you restart a server?)

tafia commented:
[clabot:check]

tafia commented:
The tests with circular dependencies are really not easy to work with.
I'll stop for today. I have commited the changes with more generics. I have also modified the `HashDb::Get` to return a borrowed value instead of an owned one. I found so far one instance where is it beneficial ([here](https://github.com/paritytech/parity-common/pull/29/commits/2e909a7ffb226f88429d3a9e8c3584d348a6630a#diff-67dc8338b0f2bb0e177c6882a4474ae2R106)) I'll probably find other ones.

Anyway, here are the current benchcmps (still using DBValue = ElasticArray)
```
$ cargo benchcmp master_bench vec_bench
 name                        master_bench ns/iter  vec_bench ns/iter  diff ns/iter  diff %  speedup 
 trie_insertions_32_mir_1k   3,531,682             3,501,244               -30,438  -0.86%   x 1.01 
 trie_insertions_32_ran_1k   3,495,609             3,459,682               -35,927  -1.03%   x 1.01 
 trie_insertions_random_mid  2,661,568             2,645,017               -16,551  -0.62%   x 1.01 
 trie_insertions_six_high    2,694,820             2,613,699               -81,121  -3.01%   x 1.03 
 trie_insertions_six_low     5,148,732             5,144,249                -4,483  -0.09%   x 1.00 
 trie_insertions_six_mid     3,453,874             3,439,671               -14,203  -0.41%   x 1.00 
 trie_iter                   2,631,114             2,415,504              -215,610  -8.19%   x 1.09 
```

tafia commented:
@debris there are many issues with the formatting. I plan to have a pass at a later stage because it is very much in flux and I cannot reliably use rustfmt on the code (too many changes). Sorry about that.

debris commented:
> @debris there are many issues with the formatting. I plan to have a pass at a later stage because it is very much in flux and I cannot reliably use rustfmt on the code (too many changes). Sorry about that.

Just make sure your code editor uses `.editorconfig`. oh, wait... `.editorconfig` is actually missing from this repo. I'll add it

dvdplm commented on test-support/patricia-trie-ethereum/src/rlp_node_codec.rs:
```diff
@@ -72,20 +72,20 @@ impl NodeCodec<KeccakHasher> for RlpNodeCodec<KeccakHasher> {
         stream.append(&partial);
         stream.append(&value);
-		stream.drain()
+        stream.drain()
```
Most definitely my fault. Apologies @tafia !

debris commented:
> Benches don't compile. I have tried to add ethereum-types = { version = "0.3.3", git = "https://github.com/paritytech/primitives" } as dev-dependencies but I got a lot of errors.

@tafia use just one of them `version =` and `git =` are contradictory

arkpar commented:
Would be nice if someone could run the benchmark on windows. System allocator on windows is significantly slower than jemalloc that is used on *nix by default.

niklasad1 commented:
@tafia 

The master branch is currently broken but if you use ```--no-default-features``` it should work (I think).
Otherwise, we will fix soon sorry!

dvdplm commented:
Fixed the triehash benchmarks [here](https://github.com/paritytech/parity-common/pull/35)

tafia commented:
> Fixed the triehash benchmarks here

Thanks! Here are the (good) results for the benches:

```
$ cargo benchcmp master vec
 name                            master ns/iter  vec ns/iter  diff ns/iter   diff %  speedup 
 triehash_insertions_32_mir_1k   2,892,706       2,773,620        -119,086   -4.12%   x 1.04 
 triehash_insertions_32_ran_1k   2,887,924       2,735,877        -152,047   -5.26%   x 1.06 
 triehash_insertions_random_mid  1,474,901       1,335,482        -139,419   -9.45%   x 1.10 
 triehash_insertions_six_high    1,802,946       1,638,116        -164,830   -9.14%   x 1.10 
 triehash_insertions_six_low     2,378,482       2,103,125        -275,357  -11.58%   x 1.13 
 triehash_insertions_six_mid     1,965,344       1,754,439        -210,905  -10.73%   x 1.12
```

Again I cannot tell if the benches really capture real use cases or not. It would be good to run it on windows as well.

tafia commented:
There are more things to be done/tested but I believe this PR is good to go already (if all the tests pass).

I think the work or reducing the usage of elastic-array should continue, again less for speed boost than for reducing unsafe.

If you have some time, please try the benches on your side too (in particular someone with a windows machine) to be certain there is no regression.

Here are all the benches (with the fix from #35):
```sh
$ cargo +nightly  bench --all > master
$ git checkout vec
$ cargo +nightly  bench --all > vec
$ cargo benchcmp master vec
 name                                 master ns/iter     vec ns/iter        diff ns/iter   diff %  speedup 
 bench_decode_nested_empty_lists      361                369                           8    2.22%   x 0.98 
 bench_decode_u256_value              65                 66                            1    1.54%   x 0.98 
 bench_decode_u64_value               51                 52                            1    1.96%   x 0.98 
 bench_keccak_256_with_empty_input    880                715                        -165  -18.75%   x 1.23 
 bench_keccak_256_with_large_input    25,887 (158 MB/s)  21,360 (191 MB/s)        -4,527  -17.49%   x 1.21 
 bench_keccak_256_with_typical_input  749 (68 MB/s)      733 (69 MB/s)               -16   -2.14%   x 1.02 
 bench_stream_1000_empty_lists        13,728             7,694                    -6,034  -43.95%   x 1.78 
 bench_stream_nested_empty_lists      264                138                        -126  -47.73%   x 1.91 
 bench_stream_u256_value              413                332                         -81  -19.61%   x 1.24 
 bench_stream_u64_value               128                65                          -63  -49.22%   x 1.97 
 compare_to_null_embedded_in_struct   2                  2                             0    0.00%   x 1.00 
 compare_to_null_in_const             2                  1                            -1  -50.00%   x 2.00 
 contains_with_non_null_key           20                 21                            1    5.00%   x 0.95 
 contains_with_null_key               4                  4                             0    0.00%   x 1.00 
 from_fixed_array                     33,435             33,345                      -90   -0.27%   x 1.00 
 instantiation                        768                772                           4    0.52%   x 0.99 
 trie_insertions_32_mir_1k            4,081,596          3,526,486              -555,110  -13.60%   x 1.16 
 trie_insertions_32_ran_1k            5,432,165          3,574,787            -1,857,378  -34.19%   x 1.52 
 trie_insertions_random_mid           2,767,175          2,577,534              -189,641   -6.85%   x 1.07 
 trie_insertions_six_high             2,910,443          2,585,836              -324,607  -11.15%   x 1.13 
 trie_insertions_six_low              5,610,654          5,157,865              -452,789   -8.07%   x 1.09 
 trie_insertions_six_mid              3,394,414          3,341,944               -52,470   -1.55%   x 1.02 
 trie_iter                            2,572,938          2,551,388               -21,550   -0.84%   x 1.01 
 triehash_insertions_32_mir_1k        2,886,954          2,786,369              -100,585   -3.48%   x 1.04 
 triehash_insertions_32_ran_1k        2,866,430          2,757,732              -108,698   -3.79%   x 1.04 
 triehash_insertions_random_mid       1,465,016          1,335,195              -129,821   -8.86%   x 1.10 
 triehash_insertions_six_high         1,789,097          1,625,481              -163,616   -9.15%   x 1.10 
 triehash_insertions_six_low          2,347,736          2,105,459              -242,277  -10.32%   x 1.12 
 triehash_insertions_six_mid          1,930,730          1,739,807              -190,923   -9.89%   x 1.11 
 u128_mul                             15,764             15,740                      -24   -0.15%   x 1.00 
 u256_add                             39,374             39,489                      115    0.29%   x 1.00 
 u256_from_be                         10                 7                            -3  -30.00%   x 1.43 
 u256_from_le                         11                 11                            0    0.00%   x 1.00 
 u256_full_mul                        60,614             60,430                     -184   -0.30%   x 1.00 
 u256_mul                             103,172            100,575                  -2,597   -2.52%   x 1.03 
 u256_mul_small                       731                740                           9    1.23%   x 0.99 
 u256_sub                             39,434             39,379                      -55   -0.14%   x 1.00 
 u512_add                             33,469             33,733                      264    0.79%   x 0.99 
 u512_mul                             497,156            494,286                  -2,870   -0.58%   x 1.01 
 u512_mul_small                       5,316              5,333                        17    0.32%   x 1.00 
 u512_sub                             33,781             33,454                     -327   -0.97%   x 1.01 
 write_default_hasher                 471                473                           2    0.42%   x 1.00 
 write_plain_hasher                   157                159                           2    1.27%   x 0.99 
```

tafia commented:
Should be ok now. Please run the benchmarks on your side before merging.

dvdplm commented:
**NOTE**: Before merging this I'd like to publish all crates in `parity-common` and make other projects stop fetching the crates from git.

tafia commented:
I have pushed a new commit to the PR because I needed some of its changes.

I have now removed elastic-array from triehash. I have also made some improvements (avoiding allocations using either iterators or reusing the same `Vec`).

I have also rebased on master.

Benches (good) compared to current master
```
$ cargo benchcmp master vec2
 name                                 master ns/iter     vec2 ns/iter       diff ns/iter   diff %  speedup 
 bench_decode_1000_values             55,539             51,863                   -3,676   -6.62%   x 1.07 
 bench_decode_nested_empty_lists      412                366                         -46  -11.17%   x 1.13 
 bench_decode_u256_value              74                 66                           -8  -10.81%   x 1.12 
 bench_decode_u64_value               52                 54                            2    3.85%   x 0.96 
 bench_keccak_256_with_empty_input    728                717                         -11   -1.51%   x 1.02 
 bench_keccak_256_with_large_input    21,546 (190 MB/s)  21,415 (191 MB/s)          -131   -0.61%   x 1.01 
 bench_keccak_256_with_typical_input  726 (70 MB/s)      733 (69 MB/s)                 7    0.96%   x 0.99 
 bench_stream_1000_empty_lists        11,403             7,351                    -4,052  -35.53%   x 1.55 
 bench_stream_nested_empty_lists      337                136                        -201  -59.64%   x 2.48 
 bench_stream_u256_value              493                349                        -144  -29.21%   x 1.41 
 bench_stream_u64_value               149                72                          -77  -51.68%   x 2.07 
 compare_to_null_embedded_in_struct   2                  2                             0    0.00%   x 1.00 
 compare_to_null_in_const             2                  2                             0    0.00%   x 1.00 
 contains_with_non_null_key           20                 21                            1    5.00%   x 0.95 
 contains_with_null_key               4                  4                             0    0.00%   x 1.00 
 instantiation                        758                772                          14    1.85%   x 0.98 
 trie_insertions_32_mir_1k            3,716,537          3,545,322              -171,215   -4.61%   x 1.05 
 trie_insertions_32_ran_1k            3,678,674          3,605,736               -72,938   -1.98%   x 1.02 
 trie_insertions_random_mid           2,673,947          2,587,504               -86,443   -3.23%   x 1.03 
 trie_insertions_six_high             3,028,857          2,597,202              -431,655  -14.25%   x 1.17 
 trie_insertions_six_low              5,550,320          5,105,243              -445,077   -8.02%   x 1.09 
 trie_insertions_six_mid              3,479,557          3,328,909              -150,648   -4.33%   x 1.05 
 trie_iter                            2,699,675          2,376,910              -322,765  -11.96%   x 1.14 
 triehash_insertions_32_mir_1k        2,898,958          2,237,217              -661,741  -22.83%   x 1.30 
 triehash_insertions_32_ran_1k        2,867,713          2,216,454              -651,259  -22.71%   x 1.29 
 triehash_insertions_random_mid       1,473,955          1,284,393              -189,562  -12.86%   x 1.15 
 triehash_insertions_six_high         1,795,295          1,506,897              -288,398  -16.06%   x 1.19 
 triehash_insertions_six_low          2,388,282          2,049,087              -339,195  -14.20%   x 1.17 
 triehash_insertions_six_mid          1,954,390          1,660,321              -294,069  -15.05%   x 1.18 
 write_default_hasher                 499                470                         -29   -5.81%   x 1.06 
 write_plain_hasher                   157                157                           0    0.00%   x 1.00 
```

debris commented:
@tafia can you rebase this pr once again? I believe we're ready to merge it

tafia commented:
done!

tafia commented:
Awesome!
